# Problem 31 「硬貨の和」 †
イギリスでは硬貨はポンド£とペンスpがあり，一般的に流通している硬貨は以下の8種類である.

1p, 2p, 5p, 10p, 20p, 50p, £1 (100p) and £2 (200p).

以下の方法で£2を作ることが可能である．

1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p

これらの硬貨を使って£2を作る方法は何通りあるか?

In [3]:
import itertools
from collections import Counter
coins=[1,2,5,10,20,50,100,200]
c=Counter()
for n in coins:
    c[n]=0
def patterns(x):
    

Counter({1: 0, 2: 0, 5: 0, 10: 0, 20: 0, 50: 0, 100: 0, 200: 0})

# Problem 32 「パンデジタル積」 †
すべての桁に 1 から n が一度だけ使われている数をn桁の数がパンデジタル (pandigital) であるということにしよう: 例えば5桁の数 15234 は1から5のパンデジタルである.

7254 は面白い性質を持っている. 39 × 186 = 7254 と書け, 掛けられる数, 掛ける数, 積が1から9のパンデジタルとなる.

掛けられる数/掛ける数/積が1から9のパンデジタルとなるような積の総和を求めよ.

HINT: いくつかの積は, 1通り以上の掛けられる数/掛ける数/積の組み合わせを持つが1回だけ数え上げよ.

In [ ]:
'''
2桁*3桁のみ
'''



# Problem 33 「桁消去分数」 †
49/98は面白い分数である.「分子と分母からそれぞれ9を取り除くと, 49/98 = 4/8 となり, 簡単な形にすることができる」と経験の浅い数学者が誤って思い込んでしまうかもしれないからである. (方法は正しくないが，49/98の場合にはたまたま正しい約分になってしまう．)

我々は 30/50 = 3/5 のようなタイプは自明な例だとする.

このような分数のうち, 1より小さく分子・分母がともに2桁の数になるような自明でないものは, 4個ある.

その4個の分数の積が約分された形で与えられたとき, 分母の値を答えよ.

In [ ]:
'''
a/b a<b

'''

# Problem 34 「桁の階乗」 †
145は面白い数である. 1! + 4! + 5! = 1 + 24 + 120 = 145となる.

各桁の数の階乗の和が自分自身と一致するような数の和を求めよ.

注: 1! = 1 と 2! = 2 は総和に含めてはならない.

In [39]:
'''
9!*9まで調べればよい
'''
from numba import jit
import math
@jit
def facteq(x):
    fac=str(x)
    facsum=0
    for n in fac:
        facsum+=math.factorial(int(n))
    if x==facsum:
        return 1
    else:
        return 0
@jit
def main():
    eqlis=[]
    test_range=9*math.factorial(9)
    for n in range(1,test_range):
        if facteq(n)==1:
            eqlis.append(n)
    print(sum(eqlis)-3)
main()

40730


# Problem 35 「巡回素数」 †
197は巡回素数と呼ばれる. 桁を回転させたときに得られる数 197, 971, 719 が全て素数だからである.

100未満には巡回素数が13個ある: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, および97である.

100万未満の巡回素数はいくつあるか?

In [63]:
'''
二桁以上のとき
0,2,4,5,6,8を含まない
1,3,7,9の組み合わせ
4**6=4096通り
'''
import itertools 
import sympy
num=[1,3,7,9]
numlis=list(itertools.product(num,num,num))
def lis_tonum(lis):
    num=0
    for n in range(len(lis)):
        num+=lis[n]*10**(len(lis)-n-1)
    return num
primes=[lis_tonum(n) for n in numlis if sympy.isprime(lis_tonum(n))==1]
primes

[113,
 131,
 137,
 139,
 173,
 179,
 191,
 193,
 197,
 199,
 311,
 313,
 317,
 331,
 337,
 373,
 379,
 397,
 719,
 733,
 739,
 773,
 797,
 911,
 919,
 937,
 971,
 977,
 991,
 997]

# Problem 36 「二種類の基数による回文数」 †
585 = 10010010012 (2進) は10進でも2進でも回文数である.

100万未満で10進でも2進でも回文数になるような数の総和を求めよ.

(注: 先頭に0を含めて回文にすることは許されない.)

# Problem 37 「切り詰め可能素数」 †
3797は面白い性質を持っている. まずそれ自身が素数であり, 左から右に桁を除いたときに全て素数になっている (3797, 797, 97, 7). 同様に右から左に桁を除いたときも全て素数である (3797, 379, 37, 3).

右から切り詰めても左から切り詰めても素数になるような素数は11個しかない. 総和を求めよ.

注: 2, 3, 5, 7を切り詰め可能な素数とは考えない.

# 38 
192 に 1, 2, 3 を掛けてみよう.

192 × 1 = 192
192 × 2 = 384
192 × 3 = 576

積を連結することで1から9の パンデジタル数 192384576 が得られる. 192384576 を 192 と (1,2,3) の連結積と呼ぶ.

同じようにして, 9 を 1,2,3,4,5 と掛け連結することでパンデジタル数 918273645 が得られる. これは 9 と (1,2,3,4,5) との連結積である.

整数と (1,2,...,n) (n > 1) との連結積として得られる9桁のパンデジタル数の中で最大のものはいくつか?

In [1]:
def uniq(li):
    li_uniq = []
    for x in li:
        if x not in li_uniq:
            li_uniq.append(x)
    return li_uniq

def ans(x):
    while x>9182:
       
        list1=list(str(x))
        list2=list(str(2*x))
        if "0" in list1 or "0" in list2:
            x=x-1
        elif len(uniq(list1))==4 and len(uniq(list2))==5:
            for y in list1:
                if y in list2:
                    list1[:]=[]
                    list2[:]=[]
                    break
                else:
                    continue
            if len(list1)>0:
                return list1[0:4]+list2[0:5]
            else:
                x=x-1
        else:
            x=x-1
print (ans(9876))

['9', '3', '2', '7', '1', '8', '6', '5', '4']


# 39 「整数の直角三角形」 
辺の長さが {a,b,c} と整数の3つ組である直角三角形を考え, その周囲の長さを p とする. p = 120のときには3つの解が存在する:

{20,48,52}, {24,45,51}, {30,40,50}

p ≤ 1000 のとき解の数が最大になる p はいくつか?

In [11]:
'''
m>n
a=m^2-n^2
b=2mn
c=m^2+n^2
p/2=m(m+n)

'''
import math
def pitag_eq(x,y):
    x.sort()
    y.sort()
    if x[0]*y[1]==x[1]*y[0]:
        return 1
    else:
        return 0
def pitag_enum(p):
    if p%2!=0:
        return 0
    else:
        p/=2
        pitag=[]
        for n in range(2,int(p)):
            if p%n==0:
                x=n
                y=p/n-x
                if x>y and y>0:
                    pitag.append([x**2-y**2,2*x*y,x**2+y**2]) 
        return pitag
pitag_enum(120)


[[20.0, 48.0, 52.0]]

# 40 「チャンパーノウン定数」 
正の整数を順に連結して得られる以下の10進の無理数を考える:

0.123456789101112131415161718192021...
小数第12位は1である.

dnで小数第n位の数を表す. d1 × d10 × d100 × d1000 × d10000 × d100000 × d1000000 を求めよ.

In [21]:
import itertools
def d(x):
    key=list(itertools.accumulate([9,180,2700,36000,450000,5400000]))
    if x<=key[0]:
        return x
        
    for n in range(1,5):
        if x<=key[n]:
            r=(x-key[n-1])%(n+1)
            q=(x-key[n-1]-r)/(n+1)
            if r==0:
                strr=str(int(10**n+q-1))
                return strr[-1]
            else:
                strr=str(int(10**n+q))
                return strr[r-1]
    return "too big"
ans=1
for n in range(1,6):
    print(d(10**n))
    ans=ans*int(d(10**n))
print (ans) 

1
5
3
7
2
210


In [2]:
#listの積を求める手法
import functools
import operator
#functool.reduce(function,iterable)
#reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) => ((((1+2)+3)+4)+5)
#operator.mul(x,y)=>x*y =lambda x,y:x*y
result = functools.reduce(operator.mul,[n for n in range(1,6)] )
result

120